File: android/core/src/main/kotlin/com/pocketbot/core/Indicators.kt
package com.pocketbot.core

import kotlin.math.min

/**
 * Indicadores técnicos. Porta direta da versão Python já testada em
 * pocket_bot/indicators.py (mesmas fórmulas), para manter a estratégia
 * idêntica entre a VPS e o app Android.
 */
object Indicators {

    /** RSI de Wilder (média móvel exponencial dos ganhos/perdas). */
    fun rsi(closes: List<Double>, period: Int = 14): List<Double?> {
        val result = MutableList<Double?>(closes.size) { null }
        if (closes.size < 2) return result

        val alpha = 1.0 / period
        var avgGain = 0.0
        var avgLoss = 0.0
        var initialized = false

        for (i in 1 until closes.size) {
            val delta = closes[i] - closes[i - 1]
            val gain = if (delta > 0) delta else 0.0
            val loss = if (delta < 0) -delta else 0.0
            if (!initialized) {
                avgGain = gain
                avgLoss = loss
                initialized = true
            } else {
                avgGain = avgGain + alpha * (gain - avgGain)
                avgLoss = avgLoss + alpha * (loss - avgLoss)
            }
            result[i] = if (avgLoss == 0.0) {
                if (avgGain == 0.0) null else 100.0
            } else {
                val rs = avgGain / avgLoss
                100.0 - (100.0 / (1.0 + rs))
            }
        }
        return result
    }

    data class BollingerBands(val upper: Double?, val mid: Double?, val lower: Double?, val width: Double?)

    /** Bandas de Bollinger para o último ponto de uma janela de closes. */
    fun bollingerBandsAt(closes: List<Double>, index: Int, period: Int = 20, stdMult: Double = 2.0): BollingerBands {
        if (index + 1 < period) return BollingerBands(null, null, null, null)
        val window = closes.subList(index + 1 - period, index + 1)
        val mid = window.average()
        val variance = window.sumOf { (it - mid) * (it - mid) } / period
        val std = kotlin.math.sqrt(variance)
        val upper = mid + stdMult * std
        val lower = mid - stdMult * std
        val width = if (mid != 0.0) (upper - lower) / mid else null
        return BollingerBands(upper, mid, lower, width)
    }

    /**
     * Parabolic SAR (Wilder). Mesma implementação de pocket_bot/indicators.py.
     * Retorna a série de valores do SAR; close > sar indica tendência de alta.
     */
    fun psar(highs: List<Double>, lows: List<Double>, afStep: Double = 0.02, afMax: Double = 0.2): List<Double> {
        val n = highs.size
        val sar = MutableList(n) { 0.0 }
        if (n == 0) return sar

        var bull = true
        var af = afStep
        var ep = highs[0]
        sar[0] = lows[0]

        for (i in 1 until n) {
            val prevSar = sar[i - 1]
            var currSar = prevSar + af * (ep - prevSar)

            if (bull) {
                val floor1 = lows[i - 1]
                val floor2 = if (i >= 2) lows[i - 2] else lows[i - 1]
                currSar = min(currSar, min(floor1, floor2))
                if (lows[i] < currSar) {
                    bull = false
                    currSar = ep
                    ep = lows[i]
                    af = afStep
                } else if (highs[i] > ep) {
                    ep = highs[i]
                    af = min(af + afStep, afMax)
                }
            } else {
                val ceil1 = highs[i - 1]
                val ceil2 = if (i >= 2) highs[i - 2] else highs[i - 1]
                currSar = maxOf(currSar, ceil1, ceil2)
                if (highs[i] > currSar) {
                    bull = true
                    currSar = ep
                    ep = highs[i]
                    af = afStep
                } else if (lows[i] < ep) {
                    ep = lows[i]
                    af = min(af + afStep, afMax)
                }
            }
            sar[i] = currSar
        }
        return sar
    }
}

File: pocket_bot/indicators.py
"""Indicadores técnicos usados pela estratégia do bot."""
import numpy as np
import pandas as pd


def rsi(series: pd.Series, period: int = 14) -> pd.Series:
    delta = series.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)
    avg_gain = gain.ewm(alpha=1 / period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1 / period, adjust=False).mean()
    rs = avg_gain / avg_loss.replace(0, np.nan)
    return 100 - (100 / (1 + rs))


def bollinger_bands(series: pd.Series, period: int = 20, std_mult: float = 2.0):
    mid = series.rolling(period).mean()
    std = series.rolling(period).std()
    upper = mid + std_mult * std
    lower = mid - std_mult * std
    width = (upper - lower) / mid
    return upper, mid, lower, width


def psar(high: pd.Series, low: pd.Series, af_step: float = 0.02, af_max: float = 0.2) -> pd.Series:
    """Parabolic SAR (Wilder). Indicador de tendência/reversão usado como
    sinal principal pelos bots de Pocket Option mais comuns (ex.: o bot
    open-source de referência pocket_option_trading_bot usa exatamente
    este indicador). Retorna a série de valores do SAR; comparar com o
    preço de fechamento indica a tendência (fechamento > SAR = alta)."""
    high = high.reset_index(drop=True)
    low = low.reset_index(drop=True)
    n = len(high)
    sar = pd.Series(index=high.index, dtype=float)
    if n == 0:
        return sar

    bull = True
    af = af_step
    ep = high.iloc[0]
    sar.iloc[0] = low.iloc[0]

    for i in range(1, n):
        prev_sar = sar.iloc[i - 1]
        curr_sar = prev_sar + af * (ep - prev_sar)

        if bull:
            curr_sar = min(curr_sar, low.iloc[i - 1], low.iloc[i - 2] if i >= 2 else low.iloc[i - 1])
            if low.iloc[i] < curr_sar:
                bull = False
                curr_sar = ep
                ep = low.iloc[i]
                af = af_step
            elif high.iloc[i] > ep:
                ep = high.iloc[i]
                af = min(af + af_step, af_max)
        else:
            curr_sar = max(curr_sar, high.iloc[i - 1], high.iloc[i - 2] if i >= 2 else high.iloc[i - 1])
            if high.iloc[i] > curr_sar:
                bull = True
                curr_sar = ep
                ep = high.iloc[i]
                af = af_step
            elif low.iloc[i] < ep:
                ep = low.iloc[i]
                af = min(af + af_step, af_max)

        sar.iloc[i] = curr_sar

    return sar
